### DISCUSSION TITLE
`easystats` for ML

### DISCUSSION THREAD

**Discussion by pdwaggoner** [easystats/easystats#340] [Ideas]
Here is where we can more succinctly workshop an idea for approaching the ML world from an easystats perspective. Per @strengejacke 's suggestion, I wanted to just get the conversation started for a simple function idea to add to `performance` (based on `model_performance.lm`):

```{R}
# quick idea for cluster objects (kmeans, hierarchical, etc.)
model_performance.cluster <- function(d, x, metrics = "all", ...) {

  if (!requireNamespace("fpc", quietly = TRUE)) {
    stop("Package \"fpc\" needed for this function to work. Please install it.",
         call. = FALSE)
  }
  
  if (!requireNamespace("factoextra", quietly = TRUE)) {
    stop("Package \"factoextra\" needed for this function to work. Please install it.",
         call. = FALSE)
  }
  
  if(!is.data.frame(d)) {
    stop('"d" must be data frame\n',
         'You have provided an object of class: ', class(d)[1])
  }
  
# requring an "eclust" object here, because the eclust function from factoextra allows for specifying many types of clustering algorithms (e.g., diana, pam, hierarchical, kmeans, etc.)
  x <- try(methods::as(x, "eclust", strict=TRUE))
  if (!inherits(x, "eclust")){
    stop('"x" must be a eclust object',
         'You have provided an object of class: ', class(x)[1])
  } 
  
  # calculate and store validation stats for cluster model
  cluster_stats <- fpc::cluster.stats(stats::dist(d), x$cluster)
  
  if (all(metrics == "all")) {
    metrics <- c("cluster.number", "cluster.size", "average.between", "average.within", 
                 "silhouette.widths", "within.cluster.ss", "dunn", "rand", "vi")
  }
  
  out <- list()
  
  # cluster.number: number of clusters
  if ("cluster.number" %in% metrics) {
    out$cluster.number <- cluster_stats$cluster.number
  }
  
  # cluster.size: number of obs in each cluster
  if ("cluster.size" %in% metrics) {
    out$cluster.size <- cluster_stats$cluster.size
  }
  
  # average.between: average distance between clusters
  if ("average.between" %in% metrics) {
    out$average.between <- cluster_stats$average.between
  }
  
  # average.within: average distance within clusters
  if ("average.within" %in% metrics) {
    out$average.within <- cluster_stats$average.within
  }
  
  # clus.avg.silwidths: average silhouette widths
  if ("silhouette.widths" %in% metrics) {
    out$silhouette.widths <- cluster_stats$clus.avg.silwidths
  }
  
  # within.cluster.ss: within clusters sum of squares
  if ("within.cluster.ss" %in% metrics) {
    out$within.cluster.ss <- cluster_stats$within.cluster.ss
  }
  
  # dunn: Dunn index
  if ("dunn" %in% metrics) {
    out$dunn <- cluster_stats$dunn
  }
 
  # corrected.rand: (corrected) Rand index assesses cluster similarity
  if ("rand" %in% metrics) {
    out$rand <- cluster_stats$corrected.rand
  }
  
  #vi: Meila’s variation of information (VI) index assesses cluster similarity
  if ("vi" %in% metrics) {
    out$vi <- cluster_stats$vi
  }
  
  out <- as.data.frame(out)
  row.names(out) <- NULL
  out
}
```

Is this more in line with something you're considering? If not, let me know as there are many directions we could take this. 

Re: `caret`, I think it would take a bit more careful thought, as with most of the more advanced ML techniques, as there is much more behind "fitting" a model, such as tuning algorithms, creating and iterating over training and testing data sets, and so on. Its significantly more involved (though not overly arduous) than fitting a basic `lm` or `glm`, for example. Let me know your thoughts here, and we can decide how to best move forward in the ML world as it makes sense with the mission and focus of the `easyverse` (very nice, by the way ;) , love it)

**Comment by strengejacke** [comment]
I think this goes into the right direction, but do you have a working example? `kmeans()` returns no object of class `cluster`, but when I run this function in general (I renamed it for testing into `mfc()`), I get this error:

```r
mfc(mtcars, kmeans(mtcars))
#> Error in kmeans(mtcars) : 'centers' must be a number or a matrix
#> Error in mfc(mtcars, kmeans(mtcars)) : 
#>   "x" must be a eclust objectYou have provided an object of class: try-error
```


```r
mfc(mtcars, kmeans(mtcars, centers = 5))
#> Error in methods::as(x, "eclust", strict = TRUE) : 
#>   no method or default for coercing “kmeans” to “eclust”
#> Error in mfc(mtcars, kmeans(mtcars, centers = 5)) : 
#>   "x" must be a eclust objectYou have provided an object of class: try-error
```

**Comment by pdwaggoner** [comment]
@strengejacke good catch! So I was quickly troubleshooting, and am having too many difficulties with the `factoextra` package (and `eclust` specifically). So to make things a bit simpler, I have updated it to just be for objects of class `kmeans`. Here is an updated function (with updated debugging as well) and a working example that works well on my end:

```{R}
# quick idea for kmeans objects

model_performance.kmeans <- function(d, x, metrics = "all", ...) {
  
  if (!requireNamespace("fpc", quietly = TRUE)) {
    stop("Package \"fpc\" needed for this function to work. Please install it.",
         call. = FALSE)
  }
  
  if(!is.data.frame(d)) {
  } else if (!is.matrix(d)){
    stop('"d" must be data frame or matrix\n',
         'You have provided an object of class: ', class(d)[1])
  }
  
  x <- try(methods::as(x, "kmeans", strict=FALSE))
  if (!inherits(x, "kmeans")){
    stop('"x" must be a kmeans object\n',
         'You have provided an object of class: ', class(x)[1])
  } 
  
  # calculate and store validation stats for cluster model
  cluster_stats <- fpc::cluster.stats(stats::dist(d), x$cluster)
  
  if (all(metrics == "all")) {
    metrics <- c("cluster.number", "cluster.size", "average.between", "average.within", 
                 "silhouette.widths", "within.cluster.ss", "dunn", "rand", "vi")
  }

  out <- list()
  
  # cluster.number: number of clusters
  if ("cluster.number" %in% metrics) {
    out$cluster.number <- cluster_stats$cluster.number
  }
  
  # cluster.size: number of obs in each cluster
  if ("cluster.size" %in% metrics) {
    out$cluster.size <- cluster_stats$cluster.size
  }
  
  # average.between: average distance between clusters
  if ("average.between" %in% metrics) {
    out$average.between <- cluster_stats$average.between
  }
  
  # average.within: average distance within clusters
  if ("average.within" %in% metrics) {
    out$average.within <- cluster_stats$average.within
  }
  
  # clus.avg.silwidths: average silhouette widths
  if ("silhouette.widths" %in% metrics) {
    out$silhouette.widths <- cluster_stats$clus.avg.silwidths
  }
  
  # within.cluster.ss: within clusters sum of squares
  if ("within.cluster.ss" %in% metrics) {
    out$within.cluster.ss <- cluster_stats$within.cluster.ss
  }
  
  # dunn: Dunn index
  if ("dunn" %in% metrics) {
    out$dunn <- cluster_stats$dunn
  }
  
  # corrected.rand: (corrected) Rand index assesses cluster similarity
  if ("rand" %in% metrics) {
    out$rand <- cluster_stats$corrected.rand
  }
  
  #vi: Meila’s variation of information (VI) index assesses cluster similarity
  if ("vi" %in% metrics) {
    out$vi <- cluster_stats$vi
  }
  
  out <- as.data.frame(out)
  row.names(out) <- NULL
  out
}

## WORKING EXAMPLE

# Prepare Data
dta <- iris[, -5] # exclude "species" 
dta <- na.omit(dta) # omit missing obs
dta <- scale(dta) # standardize

#  load fpc for function to work
library(fpc)

# k-means example
kmeans1 <- kmeans(dta, 3)

model_performance.kmeans(dta, kmeans1, metrics = "all")
```

Note, that if you want to go this route, we may have to take individual clustering algorithms (e.g., pam clustering via the `cluster` package, and so on), resulting in a family of functions for each cluster type, e.g.: `model_performance.kmeans`, `model_performance.pam`, `model_performance.hclust`, `model_performance.clara`, etc. Thoughts here?

**Comment by strengejacke** [comment]
> Note, that if you want to go this route, we may have to take individual clustering algorithms (e.g., pam clustering via the cluster package, and so on), resulting in a family of functions for each cluster type

Yes that's fine. If the core function is identical, you can wrap it into a small helper function, which is called from the class-functions.

**Comment by DominiqueMakowski** [comment]
> Yes that's fine. If the core function is identical, you can wrap it into a small helper function, which is called from the class-functions.

Yes, the idea is to find the common "denominator" to all classes, implement this subfunction as internal, and then create exported class-specific wrappers around it

**Comment by pdwaggoner** [comment]
Thanks for the dialogue here (this is really fun, by the way). I am not sure I know how to nest functions like that, specifically:

> find the common "denominator" to all classes, implement this subfunction as internal, and then create exported class-specific wrappers around it

The only chunk within the function that would need to change is:

```{R}
  x <- try(methods::as(x, "kmeans", strict=FALSE))
  if (!inherits(x, "kmeans")){
    stop('"x" must be a kmeans object\n',
         'You have provided an object of class: ', class(x)[1])
  } 
```

From here, everything else in the function is identical. Any guidance on the "helper function" idea? Apologies as this is a bit outside of my knowledge, though I am happy to learn.

**Comment by strengejacke** [comment]
For example

```r
model_performance.kmeans <- function(d, x, metrics = "all", ...) {
  
  if (!requireNamespace("fpc", quietly = TRUE)) {
    stop("Package \"fpc\" needed for this function to work. Please install it.",
         call. = FALSE)
  }
  
  if(!is.data.frame(d)) {
  } else if (!is.matrix(d)){
    stop('"d" must be data frame or matrix\n',
         'You have provided an object of class: ', class(d)[1])
  }
  
  x <- try(methods::as(x, "kmeans", strict=FALSE))
  if (!inherits(x, "kmeans")){
    stop('"x" must be a kmeans object\n',
         'You have provided an object of class: ', class(x)[1])
  } 

  my_repeated_stuff()
}

my_repeated_stuff <- function() {
  # identical code here
}
```

Maybe with some arguments.

**Comment by pdwaggoner** [comment]
This is brilliant- thanks for the suggestion. I will take a look just a bit later on. Thanks

**Comment by pdwaggoner** [comment]
Does this make sense in the `easystats` framework?

```{R}
cluster_validation <- function(...) {
  
  # calculate and store validation stats for cluster model
  cluster_stats <- fpc::cluster.stats(stats::dist(d), x$cluster)
  
  if (all(metrics == "all")) {
    metrics <- c("cluster.number", "cluster.size", "average.between", "average.within", 
                 "silhouette.widths", "within.cluster.ss", "dunn", "rand", "vi")
  }
  
  out <- list()
  
  # cluster.number: number of clusters
  if ("cluster.number" %in% metrics) {
    out$cluster.number <- cluster_stats$cluster.number
  }
  
  # cluster.size: number of obs in each cluster
  if ("cluster.size" %in% metrics) {
    out$cluster.size <- cluster_stats$cluster.size
  }
  
  # average.between: average distance between clusters
  if ("average.between" %in% metrics) {
    out$average.between <- cluster_stats$average.between
  }
  
  # average.within: average distance within clusters
  if ("average.within" %in% metrics) {
    out$average.within <- cluster_stats$average.within
  }
  
  # clus.avg.silwidths: average silhouette widths
  if ("silhouette.widths" %in% metrics) {
    out$silhouette.widths <- cluster_stats$clus.avg.silwidths
  }
  
  # within.cluster.ss: within clusters sum of squares
  if ("within.cluster.ss" %in% metrics) {
    out$within.cluster.ss <- cluster_stats$within.cluster.ss
  }
  
  # dunn: Dunn index
  if ("dunn" %in% metrics) {
    out$dunn <- cluster_stats$dunn
  }
  
  # corrected.rand: (corrected) Rand index assesses cluster similarity
  if ("rand" %in% metrics) {
    out$rand <- cluster_stats$corrected.rand
  }
  
  #vi: Meila’s variation of information (VI) index assesses cluster similarity
  if ("vi" %in% metrics) {
    out$vi <- cluster_stats$vi
  }
  
  out <- as.data.frame(out)
  row.names(out) <- NULL
  out
}

# For k means
model_performance.kmeans <- function(d, x, metrics = "all", ...) {
  
  if (!requireNamespace("fpc", quietly = TRUE)) {
    stop("Package \"fpc\" needed for this function to work. Please install it.",
         call. = FALSE)
  }
  
  if(!is.data.frame(d)) {
  } else if (!is.matrix(d)){
    stop('"d" must be data frame or matrix\n',
         'You have provided an object of class: ', class(d)[1])
  }
  
  x <- try(methods::as(x, "kmeans", strict=FALSE))
  
  if (!inherits(x, "kmeans")){
    stop('"x" must be a kmeans object\n',
         'You have provided an object of class: ', class(x)[1])
  } 
  
  cluster_validation()
}


# For hierarchical clustering
model_performance.hclust <- function(d, x, metrics = "all", ...) {
  
  if (!requireNamespace("fpc", quietly = TRUE)) {
    stop("Package \"fpc\" needed for this function to work. Please install it.",
         call. = FALSE)
  }
  
  if(!is.data.frame(d)) {
  } else if (!is.matrix(d)){
    stop('"d" must be data frame or matrix\n',
         'You have provided an object of class: ', class(d)[1])
  }
  
  x <- try(methods::as(x, "hclust", strict=FALSE))
  
  if (!inherits(x, "hclust")){
    stop('"x" must be a hclust object\n',
         'You have provided an object of class: ', class(x)[1])
  } 
  
  cluster_validation()
}
```

**Comment by strengejacke** [comment]
@DominiqueMakowski Just revisiting this issue again... We have some cluster-stuff in _parameters_. I'm not sure if it makes sense to put parts of the code into that package, and leave some parts related to model performance here?

**Comment by pdwaggoner** [comment]
Yeah, having code in separate places is a bit odd. However, given that there is no current substantive support for ML methods (whether unsupervised like clustering or otherwise), perhaps it makes most sense as is now, and then possibly as we develop support for ML, we can revisit this and consolidate into a new easystats package, `Machine Learning`, or something like that. Not sure if that makes any sense from your perspective/vision?

**Comment by DominiqueMakowski** [comment]
I am not sure having a dedicated package for ML would conceptually be meaningful, as in the end, ML is just a fancy name for predictive modelling 😁, parts of which are taken care of by performance, parameters and all (I mean "performance" and "parameters" have a scope large enough to describe parameters and performance from regression and why not neural nets and all)

However, some aspects often used in ML, such as feature reduction/selection, could indeed require a special treatment -> https://github.com/easystats/easystats/issues/47

**Comment by pdwaggoner** [comment]
@DominiqueMakowski fair enough re: having certain packages already taking care of some things related to ML. However, I'd disagree that ML is a fancy name for predictive modeling. Though, yes, forecasting and prediction are key goals of ML, so too is classification. Further, ML is premised on a fundamentally different way of building models, from train/test splits to numeric optimization techniques, all for the goal of building effective learners based on historical patterns. See the canonical paper on this treatment, for example, https://kcir.pwr.edu.pl/~witold/ai/cacm12.pdf, for whatever it might be worth.

Yet, ultimately, from an `easystats` perspective, I agree that perhaps it doesn't make conceptual sense to build out the ML side of things so explicitly. Thanks for the dialogue!

**Comment by DominiqueMakowski** [comment]
> Further, ML is premised on a fundamentally different way of building models, from train/test splits to numeric optimization techniques

True true :)

**Comment by bwiernik** [comment]
The one major thing I would like to see that's ML-related in easystats would be easier access to train-test and k-fold cross validation workflows. This is currently quite hard to do with most of the core R statistical modeling packages. For example, there isn't a good way to compute a true out-of-sample sigma or MSE from an `lm` object. 

Some specific things to this end I would like would be:
- datawizard: functions to split a sample into folds with certain proportions, optionally with stratification on some variables
- performance: options to compute model performance statistics (sigma, R2, MSE, deviance) for a model on a new dataset
- performance: a cross_validate() function to combine the above two in one go
- modelbased: inclusion of residuals in estimate_prediction output

Thoughts?
@IndrajeetPatil @strengejacke @mattansb @DominiqueMakowski

**Comment by pdwaggoner** [comment]
> This is currently quite hard to do with most of the core R statistical modeling packages. For example, there isn't a good way to compute a true out-of-sample sigma or MSE from an lm object

As discussed in an earlier post, much of these types of ML-related tasks are covered elsewhere in the tidyverse, making an easystats version likely a duplication. For example, `yardstick`, `rsample`,  `broom`, `dplyr`, etc. help use do those could of tasks you mentioned pretty cleanly. Here's a basic example first: 

```{r}
library(tidyverse)
library(ISLR)
library(broom)
library(rsample)
library(rcfss)
library(yardstick)

Auto <- as_tibble(Auto)

set.seed(1234)

auto_split <- initial_split(data = Auto, 
                            prop = 0.8)

auto_train <- training(auto_split)
auto_test <- testing(auto_split)

auto_lm <- glm(mpg ~ horsepower, 
               data = auto_train); summary(auto_lm)

# calculate the error (MSE) for training set

(train_mse <- augment(auto_lm, 
                      newdata = auto_train) %>%
  mse(truth = mpg, 
      estimate = .fitted))

# the test set

(test_mse <- augment(auto_lm, 
                     newdata = auto_test) %>%
  mse(truth = mpg, 
      estimate = .fitted))
```

Here is a more "complex" example (vary polynomial order), but in a still clean workflow.

```{r}
# visualize each model

ggplot(Auto, aes(horsepower, mpg)) +
  geom_point(alpha = .1)


ggplot(Auto, aes(horsepower, mpg)) +
  geom_point(alpha = .1) +
  geom_smooth(aes(color = "1"),
              method = "glm",
              formula = y ~ poly(x, i = 1, raw = TRUE),
              se = FALSE) +
  geom_smooth(aes(color = "2"),
              method = "glm",
              formula = y ~ poly(x, i = 2, raw = TRUE),
              se = FALSE) +
  geom_smooth(aes(color = "3"),
              method = "glm",
              formula = y ~ poly(x, i = 3, raw = TRUE),
              se = FALSE) +
  geom_smooth(aes(color = "4"),
              method = "glm",
              formula = y ~ poly(x, i = 4, raw = TRUE),
              se = FALSE) +
  geom_smooth(aes(color = "5"),
              method = "glm",
              formula = y ~ poly(x, i = 5, raw = TRUE),
              se = FALSE) +
  scale_color_brewer(type = "qual", palette = "Dark2") +
  labs(x = "Horsepower",
       y = "MPG",
       color = "Polynomial\norder") +
  theme_minimal()

# function to train the model (training set) and evaluate the model (test set), across each polynomial regression fit
poly_results <- function(train, test, i) {

  mod <- glm(mpg ~ poly(horsepower, i, raw = TRUE), data = train)

  res <- augment(mod, 
                 newdata = test) %>%
    mse(truth = mpg, 
        estimate = .fitted)
  res
}

# function to return MSE for a unique order polynomial term
library(magrittr)

poly_mse <- function(i, train, test){
  poly_results(train, test, i) %$%
    mean(.estimate)
}

cv_mse <- tibble(terms = seq(from = 1, to = 5),
                 mse_test = map_dbl(terms, poly_mse, auto_train, auto_test))

ggplot(cv_mse, aes(terms, mse_test)) +
  geom_line() +
  labs(title = "Evaluating quadratic linear models",
       subtitle = "Using test set",
       x = "Highest-order polynomial",
       y = "Mean Squared Error") +
  theme_minimal()
```

**Comment by bwiernik** [comment]
Those examples require a pretty un-easystats like syntax and philosophy.

I'm saying that we should make it as easy to do basic cross validation as we do say bootstrapping or computing an AIC

**Comment by DominiqueMakowski** [comment]
I agree with @bwiernik.

- Regarding the test/training splitting, we already have `datawizard::data_partition(iris, group="Species")` 👍 Could be improved of course

> options to compute model performance statistics (sigma, R2, MSE, deviance) for a model on a new dataset

Would that be any different than a wrapper for 1) update(model, data=test) and 2) re-run the perf functions?

> modelbased: inclusion of residuals

have already too 

``` r
head(modelbased::estimate_prediction(lm(mpg ~ vs, data=mtcars)))
#> Model-based Prediction
#> 
#> vs   | Predicted |   SE |         95% CI | Residuals
#> ----------------------------------------------------
#> 0.00 |     16.62 | 4.71 | [ 7.01, 26.23] |     -4.38
#> 0.00 |     16.62 | 4.71 | [ 7.01, 26.23] |     -4.38
#> 1.00 |     24.56 | 4.74 | [14.87, 34.24] |      1.76
#> 1.00 |     24.56 | 4.74 | [14.87, 34.24] |      3.16
#> 0.00 |     16.62 | 4.71 | [ 7.01, 26.23] |     -2.08
#> 1.00 |     24.56 | 4.74 | [14.87, 34.24] |      6.46
#> 
#> Variable predicted: mpg
```

<sup>Created on 2021-07-31 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by DominiqueMakowski** [comment]
@pdwaggoner rather than "does this feature already exist somewhere else?", the relevant question guiding easystats efforts should be "can easystats make this feature more easy to use / intuitive / neat / smart" etc. After all, the majority of our features do exist elsewhere, we just make it smooth and easy :)

**Comment by bwiernik** [comment]
> Would that be any different than a wrapper for 1) update(model, data=test) and 2) re-run the perf functions?

Yes, we don't want to refit the model. We want to use the same parameters and estimate performance with the new data. Essentially, predict(), then computing deviance/residuals/R2/etc


So basically the above and a wrapper for cross-validation are what's missing

**Comment by pdwaggoner** [comment]
Absolutely @DominiqueMakowski and I love that about the easystats ecosystem/approach. I was just referencing the earlier conversation in #165.

**Comment by mattansb** [comment]
I agree with @pdwaggoner that there are other (popular) tools out there that do many of these you've all discussed for ML models. Can we do them better? Probably (I am not a fan of the tidymodels piped-to-death design). But I think our time would be better spent focusing on non-ML models where possible.

However, I also agree with @bwiernik that we should and can make simple out-of-(training)sample validation easy (give model, give newdata, get predictions and R2, RMSEm etc...) as these are operations that are not restricted to the ML world, and we can help popularize them here (:

**Comment by bwiernik** [comment]
That's exactly my thought!
